File: Math/PowXN.py
# LeetCode 50 - Pow(x, n)
# Time Complexity: O(log n) | Space Complexity: O(1)


class Solution:
    def myPow(self, x: float, n: int) -> float:
        return self._fast_pow(x, n) if n >= 0 else 1.0 / self._fast_pow(x, -n)

    def _fast_pow(self, base: float, exponent: int) -> float:
        answer = 1.0

        while exponent > 0:
            if exponent & 1:
                answer *= base
            base *= base
            exponent >>= 1

        return answer

File: Math/PowXN.go
// LeetCode 50 - Pow(x, n)
// Time Complexity: O(log n) | Space Complexity: O(1)
package main

func myPow(x float64, n int) float64 {
	if n >= 0 {
		return fastPow(x, int64(n))
	}
	return 1.0 / fastPow(x, -int64(n))
}

func fastPow(base float64, exponent int64) float64 {
	answer := 1.0

	for exponent > 0 {
		if exponent&1 == 1 {
			answer *= base
		}
		base *= base
		exponent >>= 1
	}

	return answer
}
